README.md
# influx-spout [![Build Status](https://travis-ci.org/jumptrading/influx-spout.svg?branch=master)](https://travis-ci.org/jumptrading/influx-spout) [![Go Report Card](https://goreportcard.com/badge/github.com/jumptrading/influx-spout)](https://goreportcard.com/report/github.com/jumptrading/influx-spout)

## Overview

influx-spout is an open source messaging system that routes and processes
[InfluxDB line protocol] metrics from agents (for example [Telegraf]) to
processing and storage backends ([InfluxDB], [Kapacitor] etc.).

Key features:

- Proven ability to handle high volumes of data (>500k points per second)
  in a production environment
- Horizontal scalability
  * Multithreaded data processing within components
  * Components can be distributed on multiple servers or a fleet of containers
- Ability to add and remove endpoints without disrupting existing data flows
- Fine-grained, regexp-based control over routing of metrics data to specific
  backends
- Sanity checking of the data stream that prevents corrupt metrics data from
  reaching backends
- Batching of outgoing data to larger chunks, making it easier for backends
  to handle high-volume dataflows
- Leverages the high-performance [NATS] messaging system

[InfluxDB]: https://www.influxdata.com/time-series-platform/influxdb/
[InfluxDB line protocol]: https://docs.influxdata.com/influxdb/v1.4/write_protocols/line_protocol_tutorial/
[Telegraf]: https://www.influxdata.com/time-series-platform/telegraf/
[NATS]: https://nats.io/
[Kapacitor]: https://www.influxdata.com/time-series-platform/kapacitor/

The following diagram shows a typical influx-spout deployment, and the
flow of data between the various components:

```
                                 +-----+              +------+
                                 | UDP |              | HTTP |
                                 +--+--+              +---+--+
                                    |                     |
                                    v                     v
                           +-----------------+   +-----------------+
                           |                 |   |                 |
                           |     Listener    |   |  HTTP Listener  |
                           |                 |   |                 |
                           +--------+--------+   +--------+--------+
                                    |                     |
                                    v                     v
 +----------------+   +-------------------------------------------------+
 |                |<--+                                                 |
 |    Filter(s)   |   |                                                 |
 |                +-->|                                                 |   +---------+
 +----------------+   |                                                 |   |         |
                      |                      NATS                       +-->| Monitor |
 +----------------+   |                                                 |   |         |
 |                |<--+                                                 |   +----+----+
 | Downsampler(s) |   |                                                 |        |
 |                +-->|                                                 |        |
 +----------------+   +-------+----------------+----------------+-------+        |
                              |                |                |                |
                              v                v                v                |
                        +----------+     +----------+     +----------+           |
                        |          |     |          |     |          |           |
                        |  Writer  |     |  Writer  |     |  Writer  |           |
                        |          |     |          |     |          |           |
                        +-----+----+     +-----+----+     +-----+----+           |
                              |                |                |                |
                              v                v                v                V
                        +----------+     +----------+     +----------+   +----------------+
                        | InfluxDB |     | InfluxDB |     | InfluxDB |   | Metrics (HTTP) |
                        +----------+     +----------+     +----------+   +----------------+
```

All the influx-spout components may be run on a single host or may be
spread across multiple hosts depending on scaling and operational
requirements.

## Building

Ensure the `GOPATH` environment is properly exported and simply run
`make` to build influx-spout. We recommend building with Go 1.9+.


## Configuration

influx-spout is a single binary which will run as any of the component
modes, depending on the configuration provided. The binary takes a
single mandatory positional argument which is the path to the [TOML]
configuration file to use:

```
influx-spout [/path/to/config.toml]
```

During startup, `/etc/influx-spout.toml` is parsed first, if it exists. The
configuration specified in the file with provided on the command line overrides
any configuration in `/etc/influx-spout.toml`. This allows configuration common
to different influx-spout components to be shared.

[TOML]: https://github.com/toml-lang/toml

### Byte Sizes

All configuration options that define a bytes size are specified using a
string and contain an optional unit suffix. These examples are all
valid ways to express the same byte size: `"2097152"`, `"2048KB"`, `"2MB"`,
`"2M"` (single char), `"2mb"` (lower case), `"2 MB"` (space before unit).

### Time Durations

All configuration options that define a time duration are specified
using a string and should use a unit suffix. These examples are all
valid ways to express the same duration: `"7200s"`, `"120m"`,
`"2h"`. Valid suffixes are "ns", "us", "ms", "s", "m" and "h".


## Modes

This section documents the various influx-spout component modes and
how to configure them.

### Listener

The listener is responsible for receiving InfluxDB line protocol measurements
on a UDP port, batching them and sending them on to a single NATS subject. In
typical deployments, a single listener will exist but it is possible to run
multiple listeners.

The supported configuration options for the listener mode follow. Defaults are
shown.


```toml
mode = "listener"  # Required

# UDP port to listen on.
port = 10001

# Address of NATS server.
nats_address = "nats://localhost:4222"

# Subject to publish received measurements on. This must be a list with one item.
nats_subject = ["influx-spout"]

# The maximum number of packets to receive before forwarding batched messages
# to the NATS server. Increasing this number reduces NATS communication overhead
# but increases latency.
batch_max_count = 10

# The maximum amount of time the listener will hold on to batched
# lines. If this age is reached, the batch is written to NATS.
batch_max_age = "5m"

# The maximum size that the listener should send at once to NATS.
# This should be no bigger than the NATS server's max_payload setting (which
# defaults to 1 MB).
batch_max_size = "1MB"

# Maximum UDP socket receive buffer size. A higher value this increases the
# peak inbound traffic the listener can handle at the cost of higher memory
# consumption.
#
# The default value is good for low to medium loads but should be increased to
# support higher receive rates.
read_buffer_size = "4MB"

# The listener will publish its own metrics to this NATS subject (for consumption
# by the monitor).
nats_subject_monitor = "influx-spout-monitor"

# The listener will publish its internal metrics to the monitor component at
# this interval.
stats_interval = "3s"

# The listener will serve Kubernetes liveness and readiness probes on this port
# at /healthz and /readyz. Set to 0 (the default) to disable probes support.
probe_port = 0

# The listener will serve Go pprof requests at this port. Set to 0 (the
# default) to disable pprof support.
pprof_port = 0
```

The listener will batch up messages until one of the limits defined by the
`batch_max_count`, `batch_max_size` or `batch_max_age` options is reached.


### HTTP Listener

The HTTP listener is like the UDP listener (above) except that it receives
measurements sent using HTTP request bodies to a `/write` endpoint on the
configured port.

The supported configuration options for the HTTP listener mode follow. Defaults
are shown.

```toml
mode = "listener_http"  # Required

# TCP port to server a HTTP server on. A single "/write" endpoint is available here.
port = 13337

# Address of NATS server.
nats_address = "nats://localhost:4222"

# Subject to publish received measurements on. This must be a list with one item.
nats_subject = ["influx-spout"]

# The maximum number of inbound HTTP writes to batch before forwarding to the
# NATS server. Increasing this number reduces NATS communication overhead but
# increases latency.
batch_max_count = 10

# The maximum amount of time the listener will hold on to batched
# lines. If this age is reached, the batch is written to NATS.
batch_max_age = "5m"

# The maximum size that the listener should send at once to NATS.
# This should be no bigger than the NATS server's max_payload setting (which
# defaults to 1 MB).
batch_max_size = "1MB"

# The HTTP listener will publish its own metrics to this NATS subject (for
# consumption by the monitor).
nats_subject_monitor = "influx-spout-monitor"

# The HTTP listener will publish its internal metrics to the monitor component
# at this interval.
stats_interval = "3s"

# The HTTP listener will serve Kubernetes liveness and readiness probes on this
# port at /healthz and /readyz. Set to 0 (the default) to disable probes support.
probe_port = 0

# The HTTP listener will serve Go pprof requests at this port. Set to 0 (the
# default) to disable pprof support.
pprof_port = 0
```


The HTTP listener will batch up messages until one of the limits defined by the
`batch_max_count`, `batch_max_size` or `batch_max_age` options is reached.

### Filter

The filter is responsible for filtering measurements published to NATS
by the listener, and forwarding them on to other NATS
subjects. Multiple filters may be run to distribute the filtering
workload across machines.

The supported configuration options for the filter mode follow. Defaults are
shown.

```toml
mode = "filter"  # Required

# Address of NATS server.
nats_address = "nats://localhost:4222"

# Subject to receive measurements from (presumably from the listener).
# This must be a list with one item.
nats_subject = ["influx-spout"]

# Measurements which do not match any rule (below) are sent to this NATS subject.
nats_subject_junkyard = "influx-spout-junk"

# The filter will publish its own metrics to this NATS subject (for consumption
# by the monitor).
nats_subject_monitor = "influx-spout-monitor"

# The maximum size that the pending buffer for the NATS subject that the filter
# is reading from may become. Measurements will be dropped if this limit is reached.
nats_max_pending_size = "200MB"

# The number of filter workers to spawn.
workers = 8

# The filter will publish its internal metrics to the monitor component at this
# interval.
stats_interval = "3s"

# The filter will serve Kubernetes liveness and readiness probes on this port at
# /healthz and /readyz. Set to 0 (the default) to disable probes support.
probe_port = 0

# The filter will serve Go pprof requests at this port. Set to 0 (the default)
# to disable pprof support.
pprof_port = 0

# Incoming metrics with timestamps ± this value from the current time will be
# rejected. Metrics with timestamps that are significantly different from previously
# written timestamps negatively impact InfluxDB performance.
max_time_delta = "10m"

# At least one rule should be defined. Rules are defined using TOML's table
# syntax. The following examples show each rule type.

[[rule]]
# "basic" rules match a measurement name exactly.
type = "basic"

# For basic rules, "match" specifies an exact measurement name.
match = "cgroup"

# Measurements matching the rule are forwarded to this subject.
subject = "measurement.cgroup"

[[rule]]
# "tags" rules match measurement tags exactly.
type = "tags"

# Each item in "tags" specifies a tag key and value pair. All tags specified
# must be present for the line to match the rule.
tags = [
  ["key1", "value1"],
  ["key2", "value2"],
]

# As above.
subject = "somewhere"

[[rule]]
# "regex" rules apply a regular expression to full measurement lines.
# Note: regex rules are significantly slower than basic and tags rules. Use
# with care.
type = "regex"

# For regex rules, "match" specifies regular expression pattern to apply.
match = "host=web.+,"

# As above.
subject = "hosts.web"


[[rule]]
# "negregex" rules apply a regular expression to full measurement lines and
# match measurements which *don't* match the regular expression.
# Note: negregex rules are significantly slower than basic rules. Use with care.
type = "negregex"

# For negregex rules, "match" specifies regular expression pattern to apply.
match = "host=web.+,"

# As above.
subject = "not-web"
```

Ordering of rules in the configuration is important. Only the first rule that
matches a given measurement is applied.


### Downsampler

The downsampler reads measurements from one or more NATS subjects,
accumlates them and emits sampled values at a configured sampling
period. This is useful for generating long term archive feeds of
measurements.

Integer and float fields are averaged over each sampling period. For
string and boolean fields, the latest value for each sampling interval
is emitted.

The input subjects to a downsampler are intended to be the outputs
from the filter.

The downsampler sends its output to NATS subjects that match the input
subjects with a suffix as configured by `downsample_suffix` (default
is `-archive`). For example, averaged measurements for an input
subject of `monitoring` would be output by the downsampler to a
NATS subject named `monitoring-archive`.

Writers should be configured to consume the downsampler output
subjects and write them to an InfluxDB instance for storage.

```toml
mode = "downsampler"  # Required

# Address of NATS server.
nats_address = "nats://localhost:4222"

# Subjects to accumulate measurements from (these should be filter outputs).
nats_subject = ["influx-spout"]

# The period to sample measurements over. Averaged measurements will be
# emitted for each period at this interval.
downsample_period = "1m"

# This suffix will be added to each input NATS subject name to form the
# output subject names.
downsample_suffix = "-archive"

# The maximum size that the pending buffer for the NATS subject that the
# downsampler is reading from may become. Measurements will be dropped if
# this limit is reached.
nats_max_pending_size = "200MB"

# The downsampler will publish its own metrics to this NATS subject (for
# consumption by the monitor).
nats_subject_monitor = "influx-spout-monitor"

# The downsampler will publish its internal metrics to the monitor component at
# this interval.
stats_interval = "3s"

# The downsampler will serve Kubernetes liveness and readiness probes on this
# port at /healthz and /readyz. Set to 0 (the default) to disable probes support.
probe_port = 0
```


### Writer

A writer is responsible for reading measurements from one or more NATS subjects,
optionally filtering them and then writing matching them to an InfluxDB
instances. A writer should be run for each InfluxDB backend that influx-spout
should send measurements to.

The supported configuration options for the writer mode follow. Defaults are
shown.

```toml
mode = "writer"  # Required

# Address of NATS server.
nats_address = "nats://localhost:4222"

# The NATS subjects to receive measurements from.
nats_subject = ["influx-spout"]

# Address of the InfluxDB instance to write to.
influxdb_address = "localhost"

# Protocol of the InfluxDB instance to write to (http/https).
influxdb_protocol = "http"

# TCP port where the InfluxDB backend can be found.
influxdb_port = 8086

# The InfluxDB database name to write to. The default value is unlikely to be
# useful. Please set to an appropriate value.
influxdb_dbname = "influx-spout-junk"

# The maximum number of message chunks to collect before a write to InfluxDB
# will be triggered. Increasing this number reduces communication overhead
# with InfluxDB but increases latency.
batch_max_count = 10

# The maximum amount of message data a writer worker may collect. If this limit
# is reached, a write to InfluxDB is performed.
#
# This should be no larger than the max-body-size option configured for the
# InfluxDB being written to. The default max-body-size value is 25MB.
batch_max_size = "10MB"

# The maximum amount of time a writer worker is allowed to hold on to collected
# data. If this limit is reached, a write to InfluxDB is performed.
batch_max_age = "5m"

# The number of writer workers to spawn.
workers = 8

# The maximum amount of time a writer will wait for an InfluxDB write to
# complete. Writes which time out will be dropped.
write_timeout = "30s"

# The maximum number of batches that failed to write to InfluxDB to track for
# retry at any given time. The oldest batch being kept for retry will be
# discarded if necessary to keep the set of batches being retried within this
# limit. Set to 0 to disable write retries.
writer_retry_batches = 1

# The amount of time between InfluxDB write retry attempts.
writer_retry_interval = "10s"

# The maximum amount of time to keep retrying a given batch.
writer_retry_timeout = "1m"

# The maximum size that the pending buffer for the NATS subject that the filter
# is reading from may become. Measurements will be dropped if this limit is reached.
nats_max_pending_size = "200MB"

# The writer will publish its own metrics to this NATS subject (for consumption
# by the monitor).
nats_subject_monitor = "influx-spout-monitor"

# The writer will publish its internal metrics to the monitor component at this
# interval.
stats_interval = "3s"

# The writer will serve Kubernetes liveness and readiness probes on this port at
# /healthz and /readyz. Set to 0 (the default) to disable probes support.
probe_port = 0

# The writer will serve Go pprof requests at this port. Set to 0 (the default)
# to disable pprof support.
pprof_port = 0

# Writers can optionally include filter rules. When filter rules are configured,
# measurements which don't match a rule will be dropped by the writer instead
# of being written to InfluxDB. Rule configuration is the same as for the
# filter component, but the rule subject should be omitted.
[[rule]]
type = "basic"
match = "cgroup"
```

A writer will batch up messages until one of the limits defined by the
`batch_max_count`, `batch_max_size` or `batch_max_age` options is
reached.

The writer also supports optional HTTP basic authentication to the
InfluxDB backend. The username and password are set using the
`INFLUXDB_USER` and `INFLUXDB_PASS` environment variables. These
options can't be set via the configuration file.

### Monitor

The monitor is responsible for collecting metrics from the other
influx-spout components and serving then over HTTP in [Prometheus data
exposition format](https://prometheus.io/docs/instrumenting/exposition_formats/). Metrics
are available at `/metrics` on the configured port.

The supported configuration options for the monitor mode follow. Defaults are
shown.

```toml
mode = "monitor"  # Required

# Address of NATS server.
nats_address = "nats://localhost:4222"

# The NATS subject used by influx-spout components to report metrics to the monitor.
# The monitor will consume and aggregate metrics sent to this subject.
nats_subject_monitor = "influx-spout-monitor"

# The TCP port where the monitor will serve Prometheus formatted metrics over HTTP.
port = 9331

# The monitor will serve Kubernetes liveness and readiness probes on this port
# at /healthz and /readyz. Set to 0 (the default) to disable probes support.
probe_port = 0

# The monitor will serve Go pprof requests at this port. Set to 0 (the default)
# to disable pprof support.
pprof_port = 0
```


## Metrics

Each influx-spout component exposes internal metrics in [Prometheus
format](https://prometheus.io/docs/instrumenting/exposition_formats/). These
metrics can be retrieved using the HTTP endpoint exposed by the
monitor component.


release-notes.md
# v2.3.0 (Unreleased)

- Migrated to Go modules
- Upgraded NATS client to 1.10 and test server to NATS 2.1.4

# v2.2.0

## Listener batch size handling

- Fail config validation if listener batch size is greater than
  1MB. NATS will only accept messages of up to 1MB in size.
- Ensure the listener will never send a batch to NATS that is bigger
  than the 1MB which NATS will accept. This was previously possible
  under some circumstances.

## InfluxDB HTTP basic auth support

influx-spout writers can now authenticate to InfluxDB backends if
required. The username and password are read from two environment
variables (`INFLUXDB_USER` and `INFLUXDB_PASS`).

## Write retries

influx-spout writers will now retry failed writes for a configurable
number of times. Retries are configured by the following new options:

* `writer_retry_batches`: The maximum number of batches that failed to
  write to InfluxDB to track for retry at any given time. The oldest retry batch will be
  discarded if necessary to keep the set of batches being retried within this
  limit. Default is 1. Set to 0 to disable write retries.
* `writer_retry_interval`: The amount of time between write retry
  attempts. Default is "10s".
* `writer_retry_timeout`: The maximum amount of time to keep retrying
  a given batch. Default is "1m"

## Enforce maximum writer batch size

When writing a batch to InfluxDB it is possible for the batch size to
exceed the maximum batch size because incoming lines are unlikely to
exactly fit inside the desired batch size. A BatchSplitter is now used
to ensure that the configured batch size is never exceeded.

This is important because InfluxDB has a maximum body size that it will accept.

## Other

- Fixed escape handling bugs in downsampler component.
- The Prometheus metrics published by each influx-spout component is
  now documented in README.md.
- InfluxDB Line Protocol unescaping is now about 4% faster.
- Batches recevied by the listener component will now have a newline
  added if it is missing.
- Fixed incorrect accounting of received batches for the UDP listener.
- Fix potential issues with batch memory growth. Batch buffers now
  grow by at least the maximum UDP packet size to ensure that there is
  always sufficient memory available.

# v2.1.0

## New downsampler component

The downsamper is useful for creating rolled up versions of metrics
streams for long term archive.

It reads measurements from one or more NATS subjects, averaging
numeric fields over a sampling period (as set by `downsample_period`),
emitting the averaged values every sampling period. For non-numeric
fields, the last value seen for each sampling period is
emitted. Measurements are emitted on clean time boundaries regardless
of when the downsampler was started.

Downsampled lines are emitted to a NATS subject with the same name as
the input subject but with a suffix as set by the `downsample_suffix`
configuration option (default is `-archive`)

## filter: New "tags rule type

A new rule type has been introduced which allows matching against one
or more exact tags on measurement lines. The approach taken is much
faster than using regexes to achieve the same kinds of matches
(80-1000% depending on number tags to match and line size). This
approach is also much safer as matching is independent of tag order
and matches tag keys and values precisely.

## filter: Tag sorting

The filter now orders tag keys in all measurement lines passing
through it. This ensures the best performance when lines are inserted
into InfluxDB and is also required for the downsampler component.

## filter: Avoid unnecessary escaping

Unescaping of lines is now avoided where possible avoiding unnecessary
computation and memory allocation. This improves performance
significantly in many cases.

## filter: Measurement name hash caching

The hash of measurement names in lines are now cached. This speeds up
matching when there are multiple "basic" filter rules.

## Other

- Fixed invalid TOML in README
- Fix race in `probes` package tests
- Retry gnatsd startup in tests (a common reason for CI failure)
- Automatically manage probe port allocations in end-to-end test
- Generate releases out of Travis CI

# v2.0.0

## Configuration cleanup

**Compatibility**

Some significant changes to influx-spout's configuration file have
been made to improvement usability, clarity and consistency. These
changes break compatibility with the previous configuration file
format. Some configuration options have been renamed, some take a
different value type and one has been removed.

Here's a summary of the changes:

| Old Name              | New Name                | Old Format        | New Format                             |
|-----------------------|-------------------------|-------------------|----------------------------------------|
| `batch_max_mb`        | `batch_max_size`        | integer (MB)      | string, flexible units (e.g. `"20MB"`) |
| `nats_pending_max_mb` | `nats_max_pending_size` | integer (MB)      | string, flexible units (e.g. `"1GB"`)  |
| `read_buffer_bytes`   | `read_buffer_size`      | integer (bytes)   | string, flexible units (e.g. `"512K"`) |
| `batch`               | `batch_max_count`       | integer count     | (unchanged)                            |
| `listener_batch_size` | **Removed**             | -                 | use `batch_max_size` instead           |
| `batch_max_secs`      | `batch_max_age`         | integer (seconds) | string, flexible units (e.g. `"5m"`)   |
| `max_time_delta_secs` | `max_time_delta`        | integer (seconds) | string, flexible units (e.g. `"30s"`)  |
| `write_timeout_secs`  | `write_timeout`         | integer (seconds) | string, flexible units (e.g. `"10s"`)  |

## Listener batch aging

The listener component will now send its batch if the age of the batch
exceeds a configured limit, as set by the `batch_max_age` option. This
prevents batched lines from being held up if the rate of data received

config/config.go
// Copyright 2017 Jump Trading
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package config

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/BurntSushi/toml"
	"github.com/c2h5oh/datasize"
	"github.com/spf13/afero"
)

// The file here is parsed first. The config file given on the command
// line is then overlaid on top of it.
const commonFileName = "/etc/influx-spout.toml"

// MaxNATSMsgSize is the maximum size of message that NATS will accept.
const MaxNATSMsgSize = datasize.MB

// Config represents the configuration for a single influx-spout
// component.
type Config struct {
	Name                string            `toml:"name"`
	Mode                string            `toml:"mode"`
	NATSAddress         string            `toml:"nats_address"`
	NATSSubject         []string          `toml:"nats_subject"`
	NATSSubjectMonitor  string            `toml:"nats_subject_monitor"`
	NATSSubjectJunkyard string            `toml:"nats_subject_junkyard"`
	InfluxDBAddress     string            `toml:"influxdb_address"`
	InfluxDBProtocol    string            `toml:"influxdb_protocol"`
	InfluxDBPort        int               `toml:"influxdb_port"`
	InfluxDBUser        string            `toml:"-"`
	InfluxDBPass        string            `toml:"-"`
	DBName              string            `toml:"influxdb_dbname"`
	BatchMaxCount       int               `toml:"batch_max_count"`
	BatchMaxSize        datasize.ByteSize `toml:"batch_max_size"`
	BatchMaxAge         Duration          `toml:"batch_max_age"`
	BatchSplitter       bool              `toml:"batch_splitter"`
	Port                int               `toml:"port"`
	Workers             int               `toml:"workers"`
	WriteTimeout        Duration          `toml:"write_timeout"`
	WriterRetryBatches  int               `toml:"writer_retry_batches"`
	WriterRetryInterval Duration          `toml:"writer_retry_interval"`
	WriterRetryTimeout  Duration          `toml:"writer_retry_timeout"`
	ReadBufferSize      datasize.ByteSize `toml:"read_buffer_size"`
	NATSMaxPendingSize  datasize.ByteSize `toml:"nats_max_pending_size"`
	Rule                []Rule            `toml:"rule"`
	MaxTimeDelta        Duration          `toml:"max_time_delta"`
	StatsInterval       Duration          `toml:"stats_interval"`
	ProbePort           int               `toml:"probe_port"`
	PprofPort           int               `toml:"pprof_port"`
	Debug               bool              `toml:"debug"`
	DownsamplePeriod    Duration          `toml:"downsample_period"`
	DownsampleSuffix    string            `toml:"downsample_suffix"`
	QueueGroup          string            `toml:"queue_group"`
}

// Rule contains the configuration for a single filter rule.
type Rule struct {
	Rtype   string     `toml:"type"`
	Match   string     `toml:"match"`
	Tags    [][]string `toml:"tags"`
	Subject string     `toml:"subject"`
}

func newDefaultConfig() *Config {
	return &Config{
		NATSAddress:         "nats://localhost:4222",
		NATSSubject:         []string{"influx-spout"},
		NATSSubjectMonitor:  "influx-spout-monitor",
		NATSSubjectJunkyard: "influx-spout-junk",
		InfluxDBAddress:     "localhost",
		InfluxDBProtocol:    "http",
		InfluxDBPort:        8086,
		DBName:              "influx-spout-junk",
		BatchMaxCount:       10,
		BatchMaxAge:         NewDuration(5 * time.Minute),
		BatchSplitter:       true,
		Workers:             8,
		WriteTimeout:        NewDuration(30 * time.Second),
		WriterRetryBatches:  1,
		WriterRetryInterval: NewDuration(10 * time.Second),
		WriterRetryTimeout:  NewDuration(time.Minute),
		ReadBufferSize:      4 * datasize.MB,
		NATSMaxPendingSize:  200 * datasize.MB,
		MaxTimeDelta:        NewDuration(10 * time.Minute),
		DownsampleSuffix:    "-archive",
		StatsInterval:       NewDuration(3 * time.Second),
		ProbePort:           0,
		PprofPort:           0,
		QueueGroup:          "",
	}
}

// Validate returns an error if a problem is detected with the
// configuration. It checks a variety of possible configuration errors
// but these aren't exhaustive.
func (c *Config) Validate() error {
	if c.ProbePort < 0 || c.ProbePort > 65535 {
		return errors.New("probe_port out of range")
	}
	if c.PprofPort < 0 || c.PprofPort > 65535 {
		return errors.New("pprof_port out of range")
	}

	if c.Mode != "downsampler" && c.DownsamplePeriod.Duration != time.Duration(0) {
		return errors.New("downsample_period is only for downsampler mode")
	}

	switch c.Mode {
	case "listener", "listener_http":
		return c.validateListener()
	case "filter":
		return c.validateFilter()
	case "downsampler":
		return c.validateDownsampler()
	case "writer":
		return c.validateWriter()
	case "monitor":
	default:
		return fmt.Errorf("invalid mode: %s", c.Mode)
	}
	return nil
}

func (c *Config) validateListener() error {
	if c.Port < 1 || c.Port > 65535 {
		return errors.New("listener port out of range")
	}
	if len(c.NATSSubject) != 1 {
		return errors.New("listener should only use one NATS subject")
	}
	if c.BatchMaxSize > MaxNATSMsgSize {
		return fmt.Errorf("listener batch must be %s or smaller", MaxNATSMsgSize)
	}
	return nil
}

func (c *Config) validateFilter() error {
	if len(c.NATSSubject) != 1 {
		return errors.New("filter should only use one NATS subject")
	}
	for _, rule := range c.Rule {
		if rule.Rtype == "" {
			return errors.New(`rule missing "type"`)
		}
		if rule.Subject == "" {
			return errors.New(`rule missing "subject"`)
		}
		if rule.Rtype != "tags" && rule.Match == "" {
			return errors.New(`rule missing "match"`)
		}
		if rule.Rtype != "tags" && len(rule.Tags) > 0 {
			return fmt.Errorf(`%s rule cannot use "tags"`, rule.Rtype)
		}
		if rule.Rtype == "tags" {
			if rule.Match != "" {
				return errors.New(`tags rule cannot use "match"`)
			}
			if len(rule.Tags) == 0 {
				return errors.New(`tags rule missing "tags"`)
			}
			for _, tag := range rule.Tags {
				if len(tag) != 2 {
					return errors.New("each tags item must contain 2 values")
				}
			}
		}
	}
	return nil
}

func (c *Config) validateDownsampler() error {
	if c.DownsamplePeriod.Duration == time.Duration(0) {
		return errors.New("downsample_period must be set")
	}
	if c.DownsampleSuffix == "" {
		return errors.New("downsample_suffix must be set")
	}
	return nil
}

func (c *Config) validateWriter() error {
	if len(c.NATSSubject) < 1 {
		return errors.New("writer needs at least one NATS subject")
	}
	if c.InfluxDBPort < 1 || c.InfluxDBPort > 65535 {
		return errors.New("influxdb_port out of range")
	}
	if c.InfluxDBProtocol != "http" && c.InfluxDBProtocol != "https" {
		return errors.New("influxdb_protocol can only be http or https")
	}
	if c.InfluxDBUser != "" && c.InfluxDBPass == "" {
		return errors.New("$INFLUXDB_USER without $INFLUXDB_PASS")
	}
	if c.InfluxDBUser == "" && c.InfluxDBPass != "" {
		return errors.New("$INFLUXDB_PASS without $INFLUXDB_USER")
	}

	for _, rule := range c.Rule {
		if rule.Rtype == "" {
			return errors.New(`rule missing "type"`)
		}
		if rule.Match == "" {
			return errors.New(`rule missing "match"`)
		}
		if rule.Subject != "" {
			return fmt.Errorf("writer rules shouldn't have subject (%q)", rule.Subject)
		}
	}
	return nil
}

// NewConfigFromFile parses the specified configuration file and
// returns a Config.
func NewConfigFromFile(fileName string) (*Config, error) {
	conf := newDefaultConfig()
	if err := readConfig(commonFileName, conf); err != nil && !os.IsNotExist(err) {
		return nil, err
	}
	if err := readConfig(fileName, conf); err != nil {
		return nil, err
	}

	if conf.Mode == "" {
		return nil, errors.New("mode not specified in config")
	}

	// Set dynamic defaults.
	if conf.Name == "" {
		conf.Name = pathToConfigName(fileName)
	}

	if conf.BatchMaxSize == 0 {
		switch conf.Mode {
		case "listener", "listener_http":
			conf.BatchMaxSize = datasize.MB
		default:
			conf.BatchMaxSize = 10 * datasize.MB
		}
	}

	if conf.Port == 0 {
		switch conf.Mode {
		case "listener":
			conf.Port = 10001
		case "listener_http":
			conf.Port = 13337
		case "monitor":
			conf.Port = 9331
		}
	}

	if conf.Mode == "downsampler" && conf.DownsamplePeriod.IsZero() {
		conf.DownsamplePeriod.Duration = time.Minute
	}

	if conf.Mode == "writer" {
		conf.InfluxDBUser = os.Getenv("INFLUXDB_USER")
		conf.InfluxDBPass = os.Getenv("INFLUXDB_PASS")
	}

	if err := conf.Validate(); err != nil {
		return nil, err
	}

	return conf, nil
}

func readConfig(fileName string, conf *Config) error {
	f, err := Fs.Open(fileName)
	if err != nil {
		return err
	}
	defer f.Close()

	_, err = toml.DecodeReader(f, conf)
	if err != nil {
		return fmt.Errorf("%s: %v", fileName, err)
	}
	return nil
}

func pathToConfigName(path string) string {
	// Remove directory (if any)
	path = filepath.Base(path)

	// Remove the file extension (if any)
	ext := filepath.Ext(path)
	if ext == "" {
		return path
	}
	return path[:len(path)-len(ext)]
}

// Fs abstracts away filesystem access for the config package. It
// should only be modified by tests.
var Fs = afero.NewOsFs()
